### DISCUSSION TITLE
Keyboard plays only notes from control knobs

### DISCUSSION THREAD

**Discussion by Lumm71** [GeekFunkLabs/fluidpatcher#21] [Q&A] [answered]
First: Hello and sorry for my bad english. Hopefully I can explain my problem understandble :)
I installed the headless.py version of fluidpatcher on Raspberry Pi3. Everything seems to be installed well.
But I have a weird common problem:
Only the control keys (play, stop, record, up, down, left, right..) on my Maudio Keystation 61 play a sound, but not the notes keys? I tried to understand the config files and changed some but nothing happened.
Can you help me? Here is some output from aconnect and aseqdump

root@raspberrypi:~# aconnect -l
client 0: 'System' [type=kernel]
    0 'Timer           '
	Connecting To: 128:0
    1 'Announce        '
	Connecting To: 128:0
client 14: 'Midi Through' [type=kernel]
    0 'Midi Through Port-0'
	Connecting To: 128:0
client 20: 'Keystation 61 MK3' [type=kernel,card=1]
    0 'Keystation 61 MK3 MIDI 1'
	Connecting To: 128:0
    1 'Keystation 61 MK3 MIDI 2'
	Connecting To: 128:0
client 128: 'FLUID Synth (3271)' [type=user,pid=3271]
    0 'Synth input port (3271:0)'
	Connected From: 0:1, 14:0, 0:0, 20:0, 20:1

asqdump shows this, when I play the control keys :)

^Croot@raspberrypi:~# aseqdump -p 20:1
Waiting for data. Press Ctrl+C to end.
Source  Event                  Ch  Data
 20:1   Note on                 0, note 93, velocity 127
 20:1   Note off                0, note 93
 20:1   Note on                 0, note 93, velocity 127
 20:1   Note off                0, note 93
 20:1   Note on                 0, note 93, velocity 127
 20:1   Note off                0, note 93
 20:1   Note on                 0, note 99, velocity 127
 20:1   Note off                0, note 99
 20:1   Note on                 0, note 99, velocity 127
 20:1   Note off                0, note 99
 20:1   Note on                 0, note 98, velocity 127
 20:1   Note off                0, note 98
 20:1   Note on                 0, note 98, velocity 127

**Comment by albedozero** [comment, marked as answer]
It looks like your Keystation has been reprogrammed so that the buttons send notes instead of control changes, and I'm betting the keys have been reprogrammed to send notes on a channel other than channel 1. You can reprogram them individually or just do a factory reset, after which you should be able to get some sound by playing the keys. Here's a link to the manual if you don't have one: [https://www.manua.ls/m-audio/keystation-61-mk3/manual](https://www.manua.ls/m-audio/keystation-61-mk3/manual)

**Reply by Lumm71** [reply]
I did a factory reset and now the paino keys are playing!! Thx for this. But the control knobs still playing piano...
And of this I can't change the instruments...

**Comment by albedozero** [comment]
Looks from the manual like you need to change the mode of the buttons by hitting the "advanced" button, then press the "CC" key twice to reprogram the octave buttons, or "DAW" to reprogram the record/stop/play buttons.

**Comment by Lumm71** [comment]
Great and Thanks very much! I could change via advanced-button and toggling with DAW-button the control keys to MIDI-mode and it works!! Have a nice evening.

**Reply by albedozero** [reply]
Great! Have fun!

**Reply by yvmor** [reply]
Hello
headless.py works wit a Pi implementation only ?
I would like that the record/play/stop.. buttons (on a M-audio axiom) control the fluidpatcher application implemented on a Linux PC.
By default these butons send CC 20 cc 21 cc 22 ... but no reaction on the application
What should I do ?
Thank you
Yves M.

**Reply by albedozero** [reply]
The headlesspi.py script should work on a generic Linux PC, as long as you install the necessary extras with [these commands](https://github.com/albedozero/fluidpatcher#linux-debianubuntu). Can you share what MIDI messages your buttons are sending by following [this procedure](https://github.com/albedozero/fluidpatcher/blob/master/troubleshooting.md#monitor-midi-messages) in the troubleshooting doc?

Are you perhaps trying to use the desktop GUI program _fluidpatcher.pyw_? That one does not change patches in response to MIDI messages, but it could be made to.

**Reply by yvmor** [reply]
Thank you 
Amazing fast answser!
I had not tried headlesspi.py on my linux pc, only fluidpatcher.py, aseqdump -p 28 give things like :
24:0   Control change          0, controller 21, value 127    but no patch change
The program buttons (to assign messages to buttons, knobs ...) don't send anything, anyway useless with fluidpatcher !
This morning I try heaslesspi.py on linux pc, at first glance  notes and patch change from the MIDI controller (M-Audio Axiom) work! super ! but after a short while (dozen of seconds), the application shuts down, how can I avoid this?

**Reply by albedozero** [reply]
I updated the code to allow _fluidpatcher.pyw_ to respond to `patch` router rules in bank files, as described in the [bank file docs](https://github.com/albedozero/fluidpatcher/blob/master/patcher/file_formats.md#keywords). TLDR you can just add something like the following to the `router_rules` section in each of your bank files:
```
router_rules:
- {type: cc, chan: 1, par1: 21, patch: -1}
- {type: cc, chan: 1, par1: 22, patch: 1}
```

The problem with your Axiom is described in the [troubleshooting doc](https://github.com/albedozero/fluidpatcher/blob/master/troubleshooting.md#when-i-switch-patches-using-the-buttons-on-my-controller-the-Pi-shuts-down) - the transport buttons on the Axiom don't send a zero value when you release them, so the program thinks you're holding them down, which is the signal to safely shut down the Pi. In the latest version of the code you can set SHUTDOWN_BTN to a different control so this won't happen.
